# Salaries Exercise

Click [here](https://drive.google.com/file/d/1CeOKg0G1_CpECEWvze30fMdbuEspkmWD/view?usp=sharing) to download the file

**Import pandas as pd**

In [1]:
import pandas as pd

**Read Salaries.csv as a dataframe called sal**

In [2]:
data = pd.read_csv("/content/Salaries.csv")
sal = pd.DataFrame(data)
sal

Id       EmployeeName  \
0            1     NATHANIEL FORD   
1            2       GARY JIMENEZ   
2            3     ALBERT PARDINI   
3            4  CHRISTOPHER CHONG   
4            5    PATRICK GARDNER   
...        ...                ...   
148649  148650      Roy I Tillery   
148650  148651       Not provided   
148651  148652       Not provided   
148652  148653       Not provided   
148653  148654          Joe Lopez   

                                              JobTitle    BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1                      CAPTAIN III (POLICE DEPARTMENT)  155966.02   
2                      CAPTAIN III (POLICE DEPARTMENT)  212739.13   
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC   77916.00   
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)  134401.60   
...                                                ...        ...   
148649                                       Custodian       0.00   
148650                                    Not provided        NaN   
148651                                    Not provided        NaN   
148652                                    Not provided        NaN   
148653                      Counselor, Log Cabin Ranch       0.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0              0.00  400184.25       NaN  567595.43         567595.43  2011   
1         245131.88  137811.38       NaN  538909.28         538909.28  2011   
2         106088.18   16452.60       NaN  335279.91         335279.91  2011   
3          56120.71  198306.90       NaN  332343.61         332343.61  2011   
4           9737.00  182234.59       NaN  326373.19         326373.19  2011   
...             ...        ...       ...        ...               ...   ...   
148649         0.00       0.00       0.0       0.00              0.00  2014   
148650          NaN        NaN       NaN       0.00              0.00  2014   
148651          NaN        NaN       NaN       0.00              0.00  2014   
148652          NaN        NaN       NaN       0.00              0.00  2014   
148653         0.00    -618.13       0.0    -618.13           -618.13  2014   

        Notes         Agency  Status  
0         NaN  San Francisco     NaN  
1         NaN  San Francisco     NaN  
2         NaN  San Francisco     NaN  
3         NaN  San Francisco     NaN  
4         NaN  San Francisco     NaN  
...       ...            ...     ...  
148649    NaN  San Francisco     NaN  
148650    NaN  San Francisco     NaN  
148651    NaN  San Francisco     NaN  
148652    NaN  San Francisco     NaN  
148653    NaN  San Francisco     NaN  

[148654 rows x 13 columns]

**Check the head of the DataFrame**

In [3]:
sal.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.00  400184.25       NaN  567595.43         567595.43   
1  155966.02    245131.88  137811.38       NaN  538909.28         538909.28   
2  212739.13    106088.18   16452.60       NaN  335279.91         335279.91   
3   77916.00     56120.71  198306.90       NaN  332343.61         332343.61   
4  134401.60      9737.00  182234.59       NaN  326373.19         326373.19   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN  
1  2011    NaN  San Francisco     NaN  
2  2011    NaN  San Francisco     NaN  
3  2011    NaN  San Francisco     NaN  
4  2011    NaN  San Francisco     NaN

**Use the .info() method to find out how many entries there are**

In [4]:
sal.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            0 non-null       float64
dtypes: float64(8), int64(2), object(3)
memory usage: 14.7+ MB


**Find the average BasePay ?**

Answer: 66325.44884050643

In [5]:
sal["BasePay"].mean()

66325.4488404877

**Find is the highest amount of OvertimePay in the dataset ?**

Answer: 245131.88

In [6]:
sal["OvertimePay"].max()

245131.88

**Find the job title of  JOSEPH DRISCOLL ?**

Note: Use all caps, otherwise you may get an answer that doesn't match up (there is also a lowercase Joseph Driscoll)

Answer: CAPTAIN, FIRE SUPPRESSION

In [7]:
sal[sal["EmployeeName"] == "JOSEPH DRISCOLL"]["JobTitle"].values[0]

'CAPTAIN, FIRE SUPPRESSION'

**How much does JOSEPH DRISCOLL make (including benefits)?**

Answer: 270324.91

In [8]:
sal[sal["EmployeeName"] == "JOSEPH DRISCOLL"]["TotalPayBenefits"].values[0]

270324.91

**What is the name of highest paid person (including benefits) ?**

Note: Print the entire row of data as follows

Answer:



```
1 | NATHANIEL FORD | GENERAL MANAGER - METROPOLITAN TRANSIT...
```




In [9]:
sal.loc[sal["TotalPayBenefits"] == sal.describe().loc["max", "TotalPayBenefits"]]

Id    EmployeeName                                        JobTitle  \
0   1  NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18          0.0  400184.25       NaN  567595.43         567595.43   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN

**What is the name of lowest paid person (including benefits)? Do you notice something strange about how much he or she is paid? If yes put it as a comment inside the given code block**

Note: Print the entire row of data as follows

Answer:

```
148654 |	Joe Lopez	| Counselor, Log Cabin Ranch...
```

In [10]:
sal.loc[sal["TotalPayBenefits"] == sal.describe().loc["min", "TotalPayBenefits"]]

# Do you notice something strange about how much he or she is paid?
# Yes, Joe Lopez is paid in a negative amount (-618.13)

Id EmployeeName                    JobTitle  BasePay  OvertimePay  \
148653  148654    Joe Lopez  Counselor, Log Cabin Ranch      0.0          0.0   

        OtherPay  Benefits  TotalPay  TotalPayBenefits  Year  Notes  \
148653   -618.13       0.0   -618.13           -618.13  2014    NaN   

               Agency  Status  
148653  San Francisco     NaN

**What was the average (mean) BasePay of all employees per year? (2011-2014) ?**

Answer:

```
Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64
```

In [11]:
sal.groupby("Year")["BasePay"].mean()

Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

**How many unique job titles are there ?**

Answer: 2159

In [12]:
sal["JobTitle"].nunique()

2159

**What are the top 5 most common jobs ?**

Answer:

```
Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: JobTitle, dtype: int64
```

In [13]:
sal["JobTitle"].value_counts().head(5)

JobTitle
Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: count, dtype: int64

**How many Job Titles were represented by only one person in 2013? (e.g. Job Titles with only one occurence in 2013?)**

Answer: 202

In [14]:
sal2013 = sal[sal["Year"] == 2013]
jobcount = sal2013["JobTitle"].value_counts()
len(jobcount[jobcount == 1])

202

**How many people have the word Chief in their job title ?**

Note: This is pretty tricky

Answer: 477

In [15]:
chief = sal["JobTitle"].str.contains("Chief")
len(sal[chief])

423

**Is there a correlation between length of the Job Title string and Salary ?**

In [16]:
sal["JobTitleLength"] = sal["JobTitle"].apply(len)
sal["JobTitleLength"].corr(sal["TotalPayBenefits"])

# There is no significant correlation between the length of the Job Title string and Salary
# The value is -0.037, which is closer to 0, indicating a weak negative relationship

-0.036878445932606675

# Ecommerce Purchases Exercise

Click [here](https://drive.google.com/file/d/1B27xrTYJQEzTwnWNn872yOZUljJ-WlFT/view?usp=sharing) to download the file

**Import pandas**

In [17]:
import pandas as pd

**Read in the Ecommerce Purchases csv file and set it to a DataFrame called ecom**

In [18]:
data = pd.read_csv("/content/EcommercePurchases.csv")
ecom = pd.DataFrame(data)
ecom

Address    Lot AM or PM  \
0     16629 Pace Camp Apt. 448\nAlexisborough, NE 77...  46 in       PM   
1     9374 Jasmine Spurs Suite 508\nSouth John, TN 8...  28 rn       PM   
2                      Unit 0065 Box 5052\nDPO AP 27450  94 vE       PM   
3                 7780 Julia Fords\nNew Stacy, WA 45798  36 vm       PM   
4     23012 Munoz Drive Suite 337\nNew Cynthia, TX 5...  20 IE       AM   
...                                                 ...    ...      ...   
9995      966 Castaneda Locks\nWest Juliafurt, CO 96415  92 XI       PM   
9996  832 Curtis Dam Suite 785\nNorth Edwardburgh, T...  41 JY       AM   
9997              Unit 4434 Box 6343\nDPO AE 28026-0283  74 Zh       AM   
9998               0096 English Rest\nRoystad, IA 12457  74 cL       PM   
9999     40674 Barrett Stravenue\nGrimesville, WI 79682  64 Hr       AM   

                                           Browser Info  \
0     Opera/9.56.(X11; Linux x86_64; sl-SI) Presto/2...   
1     Opera/8.93.(Windows 98; Win 9x 4.90; en-US) Pr...   
2     Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...   
3     Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_0 ...   
4     Opera/9.58.(X11; Linux x86_64; it-IT) Presto/2...   
...                                                 ...   
9995  Mozilla/5.0 (Windows NT 5.1) AppleWebKit/5352 ...   
9996  Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...   
9997  Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_7...   
9998  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_8;...   
9999  Mozilla/5.0 (X11; Linux i686; rv:1.9.5.20) Gec...   

                              Company       Credit Card CC Exp Date  \
0                     Martinez-Herman  6011929061123406       02/20   
1     Fletcher, Richards and Whitaker  3337758169645356       11/18   
2          Simpson, Williams and Pham      675957666125       08/19   
3     Williams, Marshall and Buchanan  6011578504430710       02/24   
4           Brown, Watson and Andrews  6011456623207998       10/25   
...                               ...               ...         ...   
9995                    Randall-Sloan   342945015358701       03/22   
9996         Hale, Collins and Wilson   210033169205009       07/25   
9997                     Anderson Ltd  6011539787356311       05/21   
9998                         Cook Inc   180003348082930       11/17   
9999                       Greene Inc  4139972901927273       02/19   

      CC Security Code                  CC Provider  \
0                  900                 JCB 16 digit   
1                  561                   Mastercard   
2                  699                 JCB 16 digit   
3                  384                     Discover   
4                  678  Diners Club / Carte Blanche   
...                ...                          ...   
9995               838                 JCB 15 digit   
9996               207                 JCB 16 digit   
9997                 1                VISA 16 digit   
9998               987             American Express   
9999               302                 JCB 15 digit   

                               Email                                     Job  \
0                  pdunlap@yahoo.com  Scientist, product/process development   
1                 anthony41@reed.com                       Drilling engineer   
2     amymiller@morales-harrison.com                Customer service manager   
3        brent16@olson-robinson.info                       Drilling engineer   
4        christopherwright@gmail.com                             Fine artist   
...                              ...                                     ...   
9995          iscott@wade-garner.com                              Printmaker   
9996              mary85@hotmail.com                         Energy engineer   
9997               tyler16@gmail.com                      Veterinary surgeon   
9998         elizabethmoore@reid.net                Local government officer   
9999           rachelford@vaughn.com        

**Check the head of the DataFrame.**

In [19]:
ecom.head()

Address    Lot AM or PM  \
0  16629 Pace Camp Apt. 448\nAlexisborough, NE 77...  46 in       PM   
1  9374 Jasmine Spurs Suite 508\nSouth John, TN 8...  28 rn       PM   
2                   Unit 0065 Box 5052\nDPO AP 27450  94 vE       PM   
3              7780 Julia Fords\nNew Stacy, WA 45798  36 vm       PM   
4  23012 Munoz Drive Suite 337\nNew Cynthia, TX 5...  20 IE       AM   

                                        Browser Info  \
0  Opera/9.56.(X11; Linux x86_64; sl-SI) Presto/2...   
1  Opera/8.93.(Windows 98; Win 9x 4.90; en-US) Pr...   
2  Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...   
3  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_0 ...   
4  Opera/9.58.(X11; Linux x86_64; it-IT) Presto/2...   

                           Company       Credit Card CC Exp Date  \
0                  Martinez-Herman  6011929061123406       02/20   
1  Fletcher, Richards and Whitaker  3337758169645356       11/18   
2       Simpson, Williams and Pham      675957666125       08/19   
3  Williams, Marshall and Buchanan  6011578504430710       02/24   
4        Brown, Watson and Andrews  6011456623207998       10/25   

   CC Security Code                  CC Provider  \
0               900                 JCB 16 digit   
1               561                   Mastercard   
2               699                 JCB 16 digit   
3               384                     Discover   
4               678  Diners Club / Carte Blanche   

                            Email                                     Job  \
0               pdunlap@yahoo.com  Scientist, product/process development   
1              anthony41@reed.com                       Drilling engineer   
2  amymiller@morales-harrison.com                Customer service manager   
3     brent16@olson-robinson.info                       Drilling engineer   
4     christopherwright@gmail.com                             Fine artist   

        IP Address Language  Purchase Price  
0  149.146.147.205       el           98.14  
1     15.160.41.51       fr           70.73  
2   132.207.160.22       de            0.95  
3     30.250.74.19       es           78.04  
4     24.140.33.94       es           77.82

**How many rows and columns are there ?**

Answer

```
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
Address             10000 non-null object
Lot                 10000 non-null object
AM or PM            10000 non-null object
Browser Info        10000 non-null object
Company             10000 non-null object
Credit Card         10000 non-null int64
CC Exp Date         10000 non-null object
CC Security Code    10000 non-null int64
CC Provider         10000 non-null object
Email               10000 non-null object
Job                 10000 non-null object
IP Address          10000 non-null object
Language            10000 non-null object
Purchase Price      10000 non-null float64
dtypes: float64(1), int64(2), object(11)
memory usage: 1.1+ MB
```

In [20]:
ecom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Address           10000 non-null  object 
 1   Lot               10000 non-null  object 
 2   AM or PM          10000 non-null  object 
 3   Browser Info      10000 non-null  object 
 4   Company           10000 non-null  object 
 5   Credit Card       10000 non-null  int64  
 6   CC Exp Date       10000 non-null  object 
 7   CC Security Code  10000 non-null  int64  
 8   CC Provider       10000 non-null  object 
 9   Email             10000 non-null  object 
 10  Job               10000 non-null  object 
 11  IP Address        10000 non-null  object 
 12  Language          10000 non-null  object 
 13  Purchase Price    10000 non-null  float64
dtypes: float64(1), int64(2), object(11)
memory usage: 1.1+ MB


**What is the average Purchase Price ?**

Answer: 50.34730200000025

In [21]:
ecom["Purchase Price"].mean()

50.347302

**What were the highest purchase prices ?**

Answer: 99.989999999999995

In [22]:
ecom["Purchase Price"].max()

99.99

**What were the lowest purchase prices ?**

Answer: 0

In [23]:
ecom["Purchase Price"].min()

0.0

**How many people have English 'en' as their Language of choice on the website ?**

In [24]:
ecom[ecom["Language"] == "en"]["Language"].count()

1098

**How many people have the job title of "Lawyer" ?**


In [25]:
ecom[ecom["Job"] == "Lawyer"]["Job"].count()

30

**How many people made the purchase during the AM and how many people made the purchase during PM ?**

**(Hint: Check out [value_counts()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html) )**

Answer:

```
PM    5068
AM    4932
Name: AM or PM, dtype: int64
```

In [26]:
ecom["AM or PM"].value_counts()

AM or PM
PM    5068
AM    4932
Name: count, dtype: int64

**What are the 5 most common Job Titles ?**

Answer:

```
Interior and spatial designer    31
Lawyer                           30
Social researcher                28
Purchasing manager               27
Designer, jewellery              27
Name: Job, dtype: int64
```



In [27]:
ecom["Job"].value_counts().head(5)

Job
Interior and spatial designer    31
Lawyer                           30
Social researcher                28
Purchasing manager               27
Designer, jewellery              27
Name: count, dtype: int64

**Someone made a purchase that came from Lot: "90 WT" , what was the Purchase Price for this transaction ?**

Answer:
```
513    75.1
Name: Purchase Price, dtype: float64
```



In [28]:
ecom[ecom["Lot"] == "90 WT"]["Purchase Price"]

513    75.1
Name: Purchase Price, dtype: float64

**What is the email of the person with the following Credit Card Number: 4926535242672853**

Answer:

```
1234    bondellen@williams-garza.com
Name: Email, dtype: object
```


In [29]:
ecom[ecom["Credit Card"] == 4926535242672853]["Email"]

1234    bondellen@williams-garza.com
Name: Email, dtype: object

**How many people have American Express as their Credit Card Provider *and* made a purchase above $95 ?**

Answer

```
Address             39
Lot                 39
AM or PM            39
Browser Info        39
Company             39
Credit Card         39
CC Exp Date         39
CC Security Code    39
CC Provider         39
Email               39
Job                 39
IP Address          39
Language            39
Purchase Price      39
dtype: int64
```



In [30]:
ecom[(ecom["CC Provider"] == "American Express") & (ecom["Purchase Price"] > 95)].count()

Address             39
Lot                 39
AM or PM            39
Browser Info        39
Company             39
Credit Card         39
CC Exp Date         39
CC Security Code    39
CC Provider         39
Email               39
Job                 39
IP Address          39
Language            39
Purchase Price      39
dtype: int64

**Hard: How many people have a credit card that expires in 2025 ?**

Answer: 1033

In [31]:
ecom[ecom["CC Exp Date"].str.contains("25")]["CC Exp Date"].count()

1033

**Hard: What are the top 5 most popular email providers/hosts (e.g. gmail.com, yahoo.com, etc...)**

Answer:

```
hotmail.com     1638
yahoo.com       1616
gmail.com       1605
smith.com         42
williams.com      37
Name: Email, dtype: int64
```



In [32]:
ecom["Email"].str.split("@").str[1].value_counts().head(5)

Email
hotmail.com     1638
yahoo.com       1616
gmail.com       1605
smith.com         42
williams.com      37
Name: count, dtype: int64